### PR TITLE
Feature/threaded io

### PR DESCRIPTION
**Pull request by hahnd**
Added a new abstraction layer for accessing FSOI datastore.  Added an implementation for an S3 datastore.  Added an implementation to perform parallel IO operations.  These changes reduce the runtime of processing a request by about 20%.

### CODE REVIEW

File: python/src/fsoi/data/s3_datastore.py [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+            # create a list of S3 keys
+            contents = response['Contents']
+            sources = []
+            for item in contents:
+                sources.append({
+                    'bucket': bucket,
+                    'key': item['Key']
+                })
+
+            return sources
```
Comment (markjolah) [inline python/src/fsoi/data/s3_datastore.py]:
```
return  [ {'bucket':bucket, 'key':item['Key']} for item in response['Contents'] ]
```

File: python/src/fsoi/data/s3_datastore.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+            bucket, key = self.__to_bucket_and_key(target)
+
+            # get the S3 client
+            s3_client = self.__get_s3_client()
+
+            # check if the target exists in the S3 bucket
+            response = s3_client.head_object(Bucket=bucket, Key=key)
+
+            # check the response (successful response indicates target exists)
+            if response['ResponseMetadata']['HTTPStatusCode'] == 200:
```
Comment (markjolah) [inline python/src/fsoi/data/s3_datastore.py]:
```
return response['ResponseMetadata']['HTTPStatusCode'] == 200
```

**Comment by hahnd** [comment]
@markjolah Thanks for the comments.  I have made the changes and ready for another look.

File: python/test/test_datastore.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+    # verify that the file was copied to the data store
+    assert ds.data_exist(descriptor)
+
+    # copy from the data store to the local file
+    local_file2 = tempfile.mktemp()
+    assert ds.load_to_local_file(descriptor, local_file2)
+    with open(local_file2) as fs:
+        data2 = fs.read()
+        fs.close()
```
Comment (markjolah) [inline python/test/test_datastore.py]: Technically if you use the `with open(...) as fs:` clause, you don't need to close the file.  The `__exit__()` method of the file object will do that for you when it exits the `with` block.

**Pull request merged**